blocks/api/serializer.js

```diff
@@ -37,34 +37,45 @@ export function getSaveContent( save, attributes ) {
 }
 
 /**
- * Returns comment attributes as serialized string, determined by subset of
- * difference between actual attributes of a block and those expected based
- * on its settings.
+ * Returns attributes which ought to be saved and serialized
  *
- * @param  {Object} realAttributes     Actual block attributes
- * @param  {Object} expectedAttributes Expected block attributes
- * @return {string}                    Comment attributes
+ * When a block exists in memory it contains as its attributes
+ * both those which come from the block comment opening _and_
+ * those which come from parsing the contents of the block.
+ * Additionally they may live in a form fine for memory but
+ * which isn't valid for serialization.
+ *
+ * This function returns a filtered set of attributes which are
+ * the ones which need to be saved in order to ensure a full
+ * serialization and they are in a form which is safe to store.
+ *
+ * @param {Object<String,*>}   allAttributes Attributes from in-memory block data
+ * @param {Object<String,*>}   fromContent   Attributes which are inferred from block content
+ * @returns {Object<String,*>} filtered set of attributes for minimum safe save/serialization
  */
-export function getCommentAttributes( realAttributes, expectedAttributes ) {
-	// Find difference and build into object subset of attributes.
-	const keys = difference(
-		Object.keys( realAttributes ),
-		Object.keys( expectedAttributes )
-	);
+export function attributesToSave( allAttributes, fromContent ) {
+	// Reasons an attribute need not be saved
+	const canBeInferred = key => fromContent.hasOwnProperty( key );
+	const isUndefined = key => undefined === allAttributes[ key ];
 
-	// Serialize the comment attributes as `key="value"`.
-	return keys.reduce( ( memo, key ) => {
-		const value = realAttributes[ key ];
-		if ( undefined === value ) {
-			return memo;
-		}
+	const isValid = key => ! some( [
+		isUndefined,
+		canBeInferred,
+	], f => f( key ) );
 
-		if ( 'string' === typeof value ) {
-			return memo + `${ key }="${ value.replace( '"', '\"' ) }" `;
-		}
+	// Specific ways we need to transform the values of saved attributes
+	const escapeDoubleQuotes = value => 'string' === typeof value
+		? value.replace( '"', '\"' )
+		: value;
 
-		return memo + `${ key }="${ value }" `;
-	}, '' );
+	const transform = key => escapeDoubleQuotes( allAttributes[ key ] );
```

Comment:
fair point. the reason I did this was to avoid iterating twice. granted, that's an optimization

it did however seem reasonable to do here since this function was taking "attributes from a block" and returning "attributes that need to be saved in the comment header." if we don't transform here, we have a function which returns a set of attributes in a state that should never exist: "the group of attributes which need to serialize in the comment header but which can't be saved as-is because of potential serialization bugs"

thoughts?